Corrfunc/tests/test_theory.py

```diff
@@ -131,6 +131,99 @@ def test_narrow_extent(N, isa='fastest', nthreads=maxthreads()):
+@pytest.mark.parametrize('binref', [1, 2, 3])
+@pytest.mark.parametrize('min_sep_opt', [False, True])
+@pytest.mark.parametrize('maxcells', [1, 2, 3])
+def test_duplicate_cellpairs(autocorr, binref, min_sep_opt, maxcells,
+                             isa='fastest', nthreads=maxthreads()):
+    '''A test to implement Manodeep's example from
+    https://github.com/manodeep/Corrfunc/pull/277#issuecomment-1190921894
+    where a particle pair straddles the wrap.
+
+    Also tests the large Rmax case, where Rmax is almost as large as L/2.
+    '''
+    from Corrfunc.theory import DD
+    boxsize = 1.
+    r_bins = [0.01, 0.4]
+    pos = np.array([[0.02, 0.98],
+                    [0., 0.],
+                    [0., 0.0],
+                    ])
+
+    results = DD(autocorr, nthreads, r_bins, pos[0], pos[1], pos[2],
+                 X2=pos[0], Y2=pos[1], Z2=pos[2],
+                 boxsize=boxsize, periodic=True, isa=isa,
+                 verbose=True, max_cells_per_dim=maxcells,
+                 xbin_refine_factor=binref, ybin_refine_factor=binref,
+                 zbin_refine_factor=binref, enable_min_sep_opt=min_sep_opt,
+                 )
+
+    assert np.all(results['npairs'] == [2])
+
+    r_bins = [0.2, 0.3, 0.49]
+    pos = np.array([[0., 0.],
+                    [0., 0.],
+                    [0., 0.48],
+                    ])
+
+    results = DD(autocorr, nthreads, r_bins, pos[0], pos[1], pos[2],
+                 X2=pos[0], Y2=pos[1], Z2=pos[2],
+                 boxsize=boxsize, periodic=True, isa=isa,
+                 verbose=True, max_cells_per_dim=maxcells,
+                 xbin_refine_factor=binref, ybin_refine_factor=binref,
+                 zbin_refine_factor=binref, enable_min_sep_opt=min_sep_opt,
+                 )
+
+    assert np.all(results['npairs'] == [0, 2])
+
+
+@pytest.mark.parametrize('autocorr', [0, 1], ids=['cross', 'auto'])
+@pytest.mark.parametrize('binref', [1, 2, 3])
+@pytest.mark.parametrize('min_sep_opt', [False, True])
+@pytest.mark.parametrize('maxcells', [1, 2, 3])
+def test_rmax_against_brute(autocorr, binref, min_sep_opt, maxcells,
+                            isa='fastest', nthreads=maxthreads()):
+    '''Generate two small point clouds near particles near (0,0,0)
+    and (L/2,L/2,L/2) and compare against the brute-force answer.
+
+    Use close to the max allowable Rmax, 0.49*Lbox.
+    '''
+    np.random.seed(1234)
+    npts = 100
+    eps = 0.2  # as a fraction of boxsize
+    boxsize = 123.
+    bins = np.linspace(0.01, 0.49*boxsize, 20)
+
+    # two clouds of width eps*boxsize
+    pos = np.random.uniform(low=-eps, high=eps,
+                            size=(3, npts))*boxsize
+    pos[:, npts//2:] += boxsize/2.  # second cloud is in center of box
+    pos %= boxsize
+
+    pdiff = np.abs(pos[:, None] - pos[:, :, None])
+    pdiff[pdiff >= boxsize/2] -= boxsize
+    pdiff = (pdiff**2).sum(axis=0).reshape(-1)
+    brutecounts, _ = np.histogram(pdiff, bins=bins**2)
+    print(brutecounts)
```

Comment:
Generally speaking, I am opposed to "clever" test code. What's the performance hit if you write out a stupid `for` loop?

Reply:
I'm sure the performance is fine for 100 points, but it's way more lines of code! But more importantly, we're using numpy, and I think we should do things the "numpy way". Python `for` loops over array indices are definitely not that!

Reply:
Sure thing but then do you mind adding explicit comments to these lines, esp. 204 and 206 - and what the shapes are of each intermediate and result array. Particularly, if there is any specific assumption about numpy multi-dim indexing, it would be good to note that too.

Reply:
Yes, done!